common/jack/midiport.h

```diff
@@ -0,0 +1,170 @@
+/*
+    Copyright (C) 2004 Ian Esten
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+*/
+
+
+#ifndef __JACK_MIDIPORT_H
+#define __JACK_MIDIPORT_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <jack/weakmacros.h>
+#include <jack/types.h>
+#include <jack/eventport.h>
+#include <stdlib.h>
+
+
+/** Type for raw event data contained in @ref jack_midi_event_t. */
+typedef jack_event_data_t jack_midi_data_t;
+
+
+/** A Jack MIDI event. */
+typedef jack_event_t jack_midi_event_t;
```

Comment:
This needs to be the other way 'round.
For ABI compatibility the library needs to expose the `jack_midi_event_t` symbol.

Reply:
Aw, crap. Are you sure there isn't a way to make this work? Do I have to redo all my renaming work or only the header files in `common/jack`?

Reply:
yes indeed. having the macros makes it API compatible, but not ABI.
the MIDI functions need to be actually defined. they will be just calling the event API, but they need to be there for visibility and backwards compatibility

Reply:
You could write wrapper functions and export those.

Reply:
This implies that `jack/eventport.h` must include `jack/midiport.h`. So we are going to be stuck with the old headers always included until the next major version bump :(

Reply:
It is the other way around. `midiport.h` needs to include `eventport.h` and define the types in a way that keeps backwards compatibility.
clients built against new jack need to work with older jack and vice-versa, the exception being if a client uses the new API not available on the older versions (obviously)

Reply:
The "vice-versa" part in particular. say you have an old, closed-source JACK app. You cannot re-compile it. That still needs to work with modern JACK.

That app might be bitwig, or might be baudline,..

JACK's ABI has been stable since it's first release and is not versioned (not even major versions break the binary interface).

Reply:
I may need help on this one, I'm lost in the C header hell with obscure compiler messages